crates/typst/src/layout/grid/layout.rs

```diff
@@ -198,44 +317,156 @@ impl CellGrid {
+        // is zero when 'items.len()' is already a multiple of 'c' (thus
+        // 'items.len() % c' would be zero).
+        let items = items.into_iter();
+        let Some(item_count) = items.len().checked_add((c - items.len() % c) % c) else {
+            bail!(span, "too many cells or lines were given")
         };
-        let mut resolved_cells: Vec<Option<Entry>> = Vec::with_capacity(cell_count);
-        for cell in cells.iter().cloned() {
+        let mut resolved_cells: Vec<Option<Entry>> = Vec::with_capacity(item_count);
+        for item in items {
+            let cell = match item {
+                GridItem::HLine { y, start, end, stroke, span, position } => {
+                    let y = y.as_custom().unwrap_or_else(|| {
+                        // When no 'y' is specified for the hline, we place it
+                        // under the latest automatically positioned cell.
+                        // The current value of the auto index is always the
+                        // index of the latest automatically positioned cell
+                        // placed plus one (that's what we do in
+                        // 'resolve_cell_position'), so we subtract 1 to get
+                        // that cell's index, and place the hline below its
+                        // row. The exception is when the auto_index is 0,
+                        // meaning no automatically positioned cell was placed
+                        // yet. In that case, we place the hline at the top of
+                        // the table.
+                        auto_index
+                            .checked_sub(1)
+                            .map_or(0, |last_auto_index| last_auto_index / c + 1)
+                    });
+                    if end.is_some_and(|end| end.get() < start) {
+                        bail!(span, "line cannot end before it starts");
+                    }
+                    let line = Line { index: y, start, end, stroke, position };
+                    // Since the amount of rows is dynamic, delay placing
+                    // hlines until after all cells were placed so we can
+                    // properly verify if they are valid. Note that we can't
+                    // place hlines even if we already know they would be in a
+                    // valid row, since it's possible that we pushed pending
+                    // hlines in the same row as this one in previous
+                    // iterations, and we need to ensure that hlines from
+                    // previous iterations are pushed to the final vector of
+                    // hlines first - the order of hlines must be kept, as this
+                    // matters when determining which one "wins" in case of
+                    // conflict. Pushing the current hline before we push
+                    // pending hlines later would change their order!
+                    pending_hlines.push((span, line));
```

Comment:
I mean, I'd be okay with having a `pending_vlines` too (or even a single `pending_lines` for both), it just felt a bit unnecessary since we can do things directly in the loop for vlines. But I can change it to that if you feel like the code readability gain is worth it.

Actually `pending_hlines` isn't _strictly_ necessary, but removing it makes checking for hline errors harder since, as you mention, we don't know the number of rows yet, and you can place an hline anywhere you want regardless of the current amount of rows, in principle. So you could do something like `table.hline(y: 9392492394)`, use all of system memory by creating that many vectors inside the `hlines` vector of vectors, only for that to be actually invalid. Felt like we could do better here. So I preferred this two-pass approach for hlines specifically. (But again, it can certainly be extended to vlines too for consistency.)

Reply:
I think consistency with `pending_vlines` would be preferrable _unless_ there is some way to factor things out into a function. Then, it would probably already be understandable enough through that.

Reply:
I've created `pending_vlines` now, and also simplified the code a bit.